### DISCUSSION TITLE
Groups handling

### DISCUSSION THREAD

**Discussion by ghost** [qtile/qtile#2345] [Q&A]
Hi guys,

I have been messing around with qtile on my arch for several weeks and everithing is going great a part from 2 aspects:

1. when I launch an app which remains open in background after I have closed it (e.g. discord) it will be always open in that group. I would like instead to open it in the group where I am when I launch it in that moment, which could be a different one from the first. Is it possible?
2. when I try to share my screen using for instance discord or google meet, I can only share those applications which are in the same group of the browser. Is there any method to share an app whic is open on a different group?

Thanks to everyone who will try to help me

**Comment by m-col** [comment]
Hello! Glad you like it.

1. This is a bit of an awkward one because these programs tend to keep their "current desktop" in memory and new windows will have this visible as a hint on them for the window manager to use. They don't have to do this, but choose to, so I guess it's a "feature". Anyway a solution might be to use to `client_new` hook to move any matching new windows to the current group. Check the docs/examples repo for similar hooks to help you out, otherwise let me know and i can try to put one together.
2. Interesting question, and as far as I know it is not possible. Windows on other groups are not mapped and I don't think their contents are defined. Do you know if this is possible on other WM/DEs under X? I imagine that it _is_ possible under Wayland but unfortunately it'll be a while before we have Wayland support.

**Reply by ghost** [reply]
I am really grateful for yoru suggestions, I will surely try to use them asap

As far as your question is concerned, I honestly do not know it by sure. I always "solved" that problem minimizing other windows applications to the task bar in DEs I used in the past (gnome, kde and dde) but I guess I can't do it properly in this case because, at least I guess, you can only reduce the dimension of a window, you can't make it "disappear" in the bar shown somewhere on your screen

**Reply by ghost** [reply]
> Hello! Glad you like it.
> 
> 1. This is a bit of an awkward one because these programs tend to keep their "current desktop" in memory and new windows will have this visible as a hint on them for the window manager to use. They don't have to do this, but choose to, so I guess it's a "feature". Anyway a solution might be to use to `client_new` hook to move any matching new windows to the current group. Check the docs/examples repo for similar hooks to help you out, otherwise let me know and i can try to put one together.
> 2. Interesting question, and as far as I know it is not possible. Windows on other groups are not mapped and I don't think their contents are defined. Do you know if this is possible on other WM/DEs under X? I imagine that it _is_ possible under Wayland but unfortunately it'll be a while before we have Wayland support.

I tried a few minutes ago to apply your solution using 'client_new' function but I faced a problem: I didn't manage to find out how to refer to the current group in order to do something like this:

@libqtile.hook.subscribe.client_new
def func(c):
    c.togroup("current group")

**Reply by m-col** [reply]
You can get the current group via the qtile object:
```py
from libqtile import qtile

@libqtile.hook.subscribe.client_new
def func(c):
    c.togroup(qtile.current_group)
```
Note that for the behaviour you want you'll need to also match the client against your criteria using a `config.Match` object (unless someone replies with a better solution... the matches are something I'm not super familiar with):
```py
from libqtile import qtile, config

discord = config.Match(wm_class="discord")
@libqtile.hook.subscribe.client_new
def func(c):
    if c.match(discord):
        c.togroup(qtile.current_group)
```
NB I think this is right but it's untested.

**Reply by ghost** [reply]
> You can get the current group via the qtile object:
> 
> ```python
> from libqtile import qtile
> 
> @libqtile.hook.subscribe.client_new
> def func(c):
>     c.togroup(qtile.current_group)
> ```
> 
> Note that for the behaviour you want you'll need to also match the client against your criteria using a `config.Match` object (unless someone replies with a better solution... the matches are something I'm not super familiar with):
> 
> ```python
> from libqtile import qtile, config
> 
> discord = config.Match(wm_class="discord")
> @libqtile.hook.subscribe.client_new
> def func(c):
>     if c.match(discord):
>         c.togroup(qtile.current_group)
> ```
> 
> NB I think this is right but it's untested.

Thank you so much, tomorrow morning I'll try to do that (it doesn't solve complitely my problem because I would like to apply it to every app and not only for one specific app). I will also try to find out more about the class match and to understand why I have to use them

**Reply by ghost** [reply]
Hi m-col,

I'm really sorry not to have answered in a very fast way but I tried what you have written here:

```
from libqtile import qtile, config

discord = config.Match(wm_class="discord")
@libqtile.hook.subscribe.client_new
def func(c):
    if c.match(discord):
        c.togroup(qtile.current_group)
```

and it doesn't affect my qtile functioning... have you got any other idea? I tried to look at the documentation and I copied and pasted what there was written, but it didn't work too.

I also tried to import the subscribe library and perform the override on @subscribe.client_new but it nothing happende

**Reply by m-col** [reply]
Make sure that the `wm_class` of discord is indeed just "discord". If you use `xprop` and click on the window, it will print out some information about it, including the class. Use the string there in your `Match`.

**Reply by ghost** [reply]
I've tried again but it slill doesn't work... It is surely my fault but I don't know were I'm making a mistake.

I thought one thing: given that discord was only an example because I would like that everything that I open is loaded on the current desktop, can I write something like this?

```
from libqtile.hook import subscribe

@subscribe.client_new
def func(c):
        c.togroup(qtile.current_group)
```
because in this case I wouldn't have to match anything and it should affect every single application that is loaded, right?

**Reply by m-col** [reply]
Yeah of course!

**Reply by ghost** [reply]
I tried to do as I said but it doesn't seem to have any effect on my system... I have pasted my config file on pastebin (https://pastebin.com/aCSHVAtR) if you want to check it because I don't understand what I am doing wrong

**Comment by tych0** [comment]
> This is a bit of an awkward one because these programs tend to keep their "current desktop" in memory and new windows will have this visible as a hint on them for the window manager to use.

Actually it's an x11 thing:

```
xprop | grep DESKTOP
_NET_WM_DESKTOP(CARDINAL) = 9
```

So you'd want to clear this property when re-mapping the window. But IMO that would be a bug not a feature :)

**Reply by m-col** [reply]
I guess it depends on what it's actually doing. I was thinking "closed it" means that the window was destroyed rather than just unmapped. I see the same thing happen when I `lazy.window.kill()` keepassxc, which stays running in the background but whose next window opens on it's last desktop. Maybe the window is not actually destroyed per se?

**Reply by ghost** [reply]
I confirm you that with "closed it" I meant usinng lazy.window.kill() on that window

**Reply by ghost** [reply]
> > This is a bit of an awkward one because these programs tend to keep their "current desktop" in memory and new windows will have this visible as a hint on them for the window manager to use.
> 
> Actually it's an x11 thing:
> 
> ```
> xprop | grep DESKTOP
> _NET_WM_DESKTOP(CARDINAL) = 9
> ```
> 
> So you'd want to clear this property when re-mapping the window. But IMO that would be a bug not a feature :)

Basically I should clean that variable every time I close a window in order to be able to open it on a differeent group the following time?

**Reply by tych0** [reply]
Huh, actually I think this is a regression. I'm noticing the same behavior now, and it didn't work this way previously...

**Reply by tych0** [reply]
Oh, no, ignore me. I have additional matches for my groups.
